adcircpy/forcing/tides/hamtide.py

```diff
@@ -1,105 +1,147 @@
-import os
-import pathlib
-import sys
+from functools import lru_cache
+from os import PathLike
+from pathlib import Path
 
 from netCDF4 import Dataset
 import numpy as np
 from scipy.interpolate import griddata
 
 # https://icdc.cen.uni-hamburg.de/en/hamtide.html
-base_url = 'https://icdc.cen.uni-hamburg.de/thredds/dodsC/ftpthredds/hamtide/'
+from adcircpy.forcing.tides.dataset import TidalDataset
 
 
-class HamtideResource:
-
-    def __set__(self, obj, resource):
-        if resource is None:
-            resource = {'elevation': {}, 'velocity': {}}
-            for constituent in obj.constituents:
-                for key in resource.keys():
-                    resource[key].update({constituent: None})
-        else:
-            raise NotImplementedError('Check that static files exist.')
-            _resource = pathlib.Path(resource)
-            for file in _resource.glob('*.nc'):
-                print(file)
-
-        obj.__dict__['resource'] = resource
-
-    def __get__(self, obj, val):
-        return obj.__dict__['resource']
-
-
-class HAMTIDE:
+class HAMTIDE(TidalDataset):
     '''
     Taguchi, E., D. Stammer and W. Zahel (2010), Estimation of deep ocean tidal energy dissipation based on the high-resolution data-assimilative HAMTIDE model (to be submitted to J. Geophys. Res.). 
     '''
 
-    _resource = HamtideResource()
-
-    def __init__(self, resource=None):
-        self._resource = resource
-
-    def __call__(self, constituent, vertices):
-        amp = self.get_amplitude(constituent, vertices)
-        phase = self.get_phase(constituent, vertices)
-        return (amp, phase)
+    CONSTITUENTS = ['S2', 'Q1', 'P1', 'O1', 'N2', 'M2', 'K2', 'K1']
+    OPENDAP_URL = 'https://icdc.cen.uni-hamburg.de/thredds/dodsC/ftpthredds/hamtide/'
 
-    def get_amplitude(self, constituent, vertices):
-        vertices = np.asarray(vertices)
+    def __init__(self, path: PathLike = None):
+        if path is not None:
+            path = Path(path)
+            if len(list(path.glob('*.nc'))) == 0:
```

Comment:
We need to do some app_data_dir sorcery here as well (for when len()>0)

Reply:
moved to separate issue